llpc/context/llpcCompiler.cpp

```diff
@@ -2438,6 +2488,14 @@ Result Compiler::buildRayTracingPipelineInternal(RayTracingContext &rtContext,
                                    static_cast<void *>(&helperThreadPayload));
 
     unsigned moduleIndex = 0;
+
+    std::vector<std::thread> workers(cl::AddRtHelpers);
+    for (std::thread &worker : workers) {
+      worker = std::thread([&helperThreadProvider, &helperThreadPayload] {
+        helperThreadBuildRayTracingPipelineElf(helperThreadProvider, &helperThreadPayload);
+      });
+    }
+
     while (!helperThreadPayload.helperThreadJoined && helperThreadProvider->GetNextTask(&moduleIndex)) {
```

Comment:
Don't we need to synchronize here to avoid a race condition on `helperThreadPayload.helperThreadJoined`? The threads will set `helperThreadPayload.helperThreadJoined = true` in `helperThreadBuildRayTracingPipelineElf`. Either result of the race should be fine, but it is still UB. Did you try running thread sanitizer on this change?

How is this race avoided in the existing parallel processing scheme not relying on `InternalHelperThreadProvider`? Are the threads somehow synchronized by `SetTasks` in that case?

My understanding of this is that we're also doing work in the main thread unless (until?) helper threads do work?

Reply:
Changed the volatile to atomics.